common/network-common/src/main/java/org/apache/spark/network/sasl/aes/AesEncryption.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.network.sasl.aes;
+
+import java.io.IOException;
+import java.nio.channels.WritableByteChannel;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.*;
+import io.netty.util.AbstractReferenceCounted;
+import org.apache.commons.crypto.stream.CryptoInputStream;
+import org.apache.commons.crypto.stream.CryptoOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.spark.network.util.ByteArrayReadableChannel;
+import org.apache.spark.network.util.ByteArrayWritableChannel;
+
+public class AesEncryption {
```

Comment:
I'm not a big fan of this class. It seems to be complicating the code instead of helping.

Instead, if these static nested classes were instead child classes of `AesCipher`, you could simplify the code a lot.

First, instead of adding the `negotiate` method to the SASL-related classes, it could live in the AesCipher code. e.g., following up on my comment in `SaslRpcHandler`:

```
if (!isAuthenticated) {
  // code around current L125
}

// Here goes code to finish the AES key negotiation.
AesCipher cipher = new AesCipher(...);
rpcCallback.reply(cipher.createConfigMessage(...));
cipher.addToChannel(...);
```

(Didn't really look at API details but hope you get the idea.)

Second, if things like `AesEncryptHandler` and `EncryptMessage` were nested non-static classes, you could remove the stream caches you have, making their life cycle more explicitly managed (and removing some weird code in the process).